orttraining/orttraining/training_ops/cuda/nn/dropout_impl.cu

```diff
@@ -65,5 +73,103 @@ SPECIALIZED_DROPOUT_GRAD_IMPL(float)
 SPECIALIZED_DROPOUT_GRAD_IMPL(double)
 SPECIALIZED_DROPOUT_GRAD_IMPL(half)
 
+constexpr int UNROLL = 4;
+
+template <typename T, bool has_residual>
+__global__ void BiasDropoutKernel(
+    const int64_t N,
+    const fast_divmod fdm_dim,
+    const float ratio,
+    const std::pair<uint64_t, uint64_t> seeds,
+    const T* X_data,
+    const T* bias_data,
+    const T* residual_data,
+    T* Y_data,
+    bool* mask_data) {
+  const float p = 1.0f - ratio;
+  const T scale = T(1.0f / p);
+
+  CUDA_LONG idx = blockDim.x * blockIdx.x + threadIdx.x;
+  CUDA_LONG step_size = gridDim.x * blockDim.x * UNROLL;
+  CUDA_LONG rounded_size = ((N - 1) / step_size + 1) * step_size;
+
+  curandStatePhilox4_32_10_t state;
+  curand_init(seeds.first, idx, seeds.second, &state);
+
+  // We ensure every thread generates the same number of random numbers (by rounding
+  // up the size) and at the same timestep (by syncing threads).
+  // From CUDA curand documentation:
+  //   The Philox_4x32_10 algorithm is closely tied to the thread and block count.
+  //   Each thread computes 4 random numbers in the same time thus the most efficient
+  //   use of Philox_4x32_10 is to generate a multiple of 4 times number of threads.
+  for (CUDA_LONG id = idx; id < rounded_size; id += step_size) {
+    float4 rand = curand_uniform4(&state);
+
+    for (CUDA_LONG i = 0; i < UNROLL; i++) {
+      CUDA_LONG li = id + gridDim.x * blockDim.x * i;
+      if (li < N) {
+        int index, offset;
+        fdm_dim.divmod(li, index, offset);
+        T bias = bias_data[offset];
+
+        mask_data[li] = (&rand.x)[i] < p;
+        T output_data = (X_data[li] + bias) * T(mask_data[li]) * scale;
+        if (has_residual) {
+          output_data += residual_data[li];
+        }
+
+        Y_data[li] = output_data;
+      }
+    }
+
+    __syncthreads();
+  }
+}
+
+template <typename T>
+void BiasDropoutKernelImpl(
+    const cudaDeviceProp& prop,
+    const int64_t N,
+    const fast_divmod fdm_dim,
+    const float ratio,
+    PhiloxGenerator& generator,
+    const T* X_data,
+    const T* bias_data,
+    const T* residual_data,
+    T* Y_data,
+    bool* mask_data) {
+  const int block_size = 256;
+  const int blocks_per_sm = prop.maxThreadsPerMultiProcessor / block_size;
+  const int grid_size = std::min(prop.multiProcessorCount * blocks_per_sm, static_cast<int>(CeilDiv(N, block_size)));
+
+  // Compute the number of random numbers generated by each thread, and increment philox generator offset by that amount.
+  const uint64_t counter_offset = static_cast<uint64_t>(((N - 1) / (block_size * grid_size * UNROLL) + 1) * UNROLL);
+  auto seeds = generator.NextPhiloxSeeds(counter_offset);
+
+  if (residual_data == nullptr) {
+    BiasDropoutKernel<T, false><<<grid_size, block_size, 0>>>(N, fdm_dim, ratio, seeds, X_data, bias_data, residual_data, Y_data, mask_data);
```

Comment:
>false [](start = 25, length = 5)

is there significant speedup from making has_residual a template parameter?

Reply:
I didn't measure this... I took it as a good practice to switch with template static variable, so compiler will optimize away the switch 

---
In reply to: [441091084](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r441091084) [](ancestors = 441091084)